lib/galaxy/tools/__init__.py

```diff
@@ -1716,6 +1716,28 @@ def exec_before_job(self, app, inp_data, out_data, param_dict={}):
     def exec_after_process(self, app, inp_data, out_data, param_dict, job=None):
         pass
 
+    def handle_outputs_when_empty(self, app, out_data, out_collections):
+        for key, value in self.outputs.items():
+            if value.when_empty:
+                data = out_data.get(key, None)
+                if data and data.get_total_size() == 0:
+                    if value.when_empty == 'hide':
+                        data.visible = False
+                        log.debug('Hiding empty output dataset: %s', data)
+                    if value.when_empty == 'delete':
+                        data.deleted = True
+                        log.debug('Deleting empty output dataset: %s', data)
+        for key, value in self.output_collections.items():
+            if value.when_empty:
+                collection = out_collections.get(key, None)
```

Comment:
If you're mapping over this tool will collection be a DC (DatasetCollection) here instead of an HDCA (HistoryDatasetCollectionAssociation)? The operations below wouldn't make sense or be possible in that context I believe. I'd strongly recommend (insist on?) adding an API test test_tools.py that maps over an example of such a tool with an output collection and clarify what the behavior should be in that context.